### PR TITLE
Support special characters for <simple_term> in CONTAINS predicate

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

This commit adds support for special characters in `<simple_term>` search conditions in the full-text search CONTAINS predicate. Previously, Babelfish threw unsupported error for the usage of special characters. This commit includes the following changes:
- Wrapping the column on which full-text search needs to be performed with the function to replace special characters with their respective unique hashes for consistent lookup
- Rewriting the search string query with extra clause to cover all the cases during lookup

Signed-off-by: Roshan Kanwar [rskanwar@amazon.com](mailto:rskanwar@amazon.com)

### Issues Resolved

Task: BABEL-4651

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** Yes

* **Arbitrary inputs -** Yes

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 1]
```diff
@@ -2590,6 +2592,27 @@ type_name(PG_FUNCTION_ARGS)
     PG_RETURN_NULL();
 }
 
+Datum
+replace_special_chars_fts(PG_FUNCTION_ARGS)
+{
+    text		*input_text = PG_GETARG_TEXT_P(0);
+    char		*input_str = text_to_cstring(input_text);
+	char		*output_str;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c:2603]: Indentation

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 2]
```diff
@@ -2590,6 +2592,27 @@ type_name(PG_FUNCTION_ARGS)
...
+
+    // Modify the input_str in place
+    output_str = replace_special_chars_FTS(input_str);
+
+    // Convert the modified input_str back to text
+    result_text = cstring_to_text(output_str);
+
+    // Free the memory allocated for input_str
+    pfree(input_str);
+	pfree(output_str);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c:2614]: Indentation

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 3]
```diff
@@ -896,6 +896,13 @@ CREATE OR REPLACE FUNCTION sys.has_dbaccess(database_name SYSNAME) RETURNS INTEG
 'babelfishpg_tsql', 'has_dbaccess'
 LANGUAGE C STABLE STRICT;
 
+-- This function performs replacing special characters to their corresponding unique hashes
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:899]: Indentation

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.0.0--4.1.0.sql [hunk 4]
```diff
@@ -2093,6 +2093,13 @@ AS 'babelfishpg_tsql', 'session_context'
 LANGUAGE C;
 GRANT EXECUTE ON FUNCTION sys.session_context TO PUBLIC;
 
+-- This function performs replacing special characters to their corresponding unique hashes
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.0.0--4.1.0.sql:2036]: Indentation

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 5]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+            while (isspace((unsigned char)*(inputPtr + 1))) {
+                // Handle multiples spaces between words and skip over additional spaces
+                inputPtr++;
+            }
+            appendStringInfoString(&output, "<->");
+        } else {
+            // Copy the character
+            appendStringInfoChar(&output, *inputPtr);
+        }
+    }
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y:182]: we are not appending '\0', please do it otherwise it assumes that data is always 0ed out during allocation

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 6]
```diff
@@ -2,6 +2,9 @@
 #include "postgres.h"
 #include "lib/stringinfo.h"
 #include <ctype.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: Ideally we should include std headers, either use PG provided definition of lib functions or include them via PG headers like "c.h"

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 7]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+    // Rewriting the query in format one<->two | oneUniqueHashtwo in order to handle special characters
+    leftStr = pstrdup(trimmedInputStr);
+    rightStr = pstrdup(trimmedInputStr);
+
+    replaceMultipleSpacesAndSpecialChars(trimmedInputStr, &leftStr, &rightStr, isEnclosedInQuotes);
+    
+    inputLength = strlen(leftStr);
+
+    initStringInfo(&output);
+    appendStringInfoString(&output, "");
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y:158]: what is the point of this statement?

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 8]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+                     errmsg("Syntax error near '%s' in the full-text search condition '%s'.", input + i + 1, input)));
+            }
+
+            if (strchr(forbiddenCharsInQuotes, input[i]) != NULL) {
+                // Replace forbiddenCharsInQuotes characters based on the position
+                if (i + 1 < inputLen && strchr(forbiddenCharsInQuotes, input[i + 1]) != NULL) {
+                    appendStringInfoString(&modifiedInput, "");
+                } else {
+                    if (i == inputLen - 1) {
+                        appendStringInfoString(&modifiedInput, "");
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: this is unnecessary

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 9]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+            // character " in between a phrase should throw syntax error
+            if (i != 0 && i != inputLen - 1 && input[i] == '"') {
+                ereport(ERROR,
+                    (errcode(ERRCODE_SYNTAX_ERROR),
+                     errmsg("Syntax error near '%s' in the full-text search condition '%s'.", input + i + 1, input)));
+            }
+
+            if (strchr(forbiddenCharsInQuotes, input[i]) != NULL) {
+                // Replace forbiddenCharsInQuotes characters based on the position
+                if (i + 1 < inputLen && strchr(forbiddenCharsInQuotes, input[i + 1]) != NULL) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: using while loop to skip all forbidden characters till last one will be simple

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 10]
```diff
@@ -18,12 +21,14 @@
 #define YYpALLOC palloc
 #define YYFREE   pfree
 
+extern char *replace_special_chars_FTS(char *input_str);
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: Nit: keep the name in lower case or camel case.

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 11]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+                    }
                 }
-                appendStringInfoString(&output, "<->");
-            } else if (*inputPtr != '-' && containsSpecialCharacters((unsigned char)*inputPtr)) {   // Handle special characters and other languages
-                pfree(trimmedInputStr);
-                yyerror(NULL, "Full-text search conditions with special characters or languages other than English are not currently supported in Babelfish");
-            } else {
-                // Copy the character
-                appendStringInfoChar(&output, *inputPtr);
+                continue;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: why not if/else

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 12]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
-            if (*inputPtr != '-' && containsSpecialCharacters((unsigned char)*inputPtr)) {   // Handle special characters and other languages
-                pfree(trimmedInputStr);
-                yyerror(NULL, "Full-text search conditions with special characters or languages other than English are not currently supported in Babelfish");
+
+        // Null-terminate the modified input string
+        appendStringInfoChar(&modifiedInput, '\0');
+
+        // Replace the input with the modified input string
+        strncpy(input, modifiedInput.data, modifiedInput.len);
+        inputLen = strlen(input);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: why to copy here? we can operate on modifiedInput itself

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 13]
```diff
@@ -136,48 +147,170 @@ char* translate_simple_term(const char* inputStr) {
...
+ * 2. special character (@) separated words
+ *
+ * Case 1: "one two" will generate "one two" and "one@two"
+ * Case 2: "one$two" will generate "one two" and "one@two"
+ */
+static void 
+replaceMultipleSpacesAndSpecialChars(char* input, char **str1, char **str2, bool isEnclosedInQuotes) {
+    size_t inputLen = strlen(input);
+    size_t i;
+    int inSpace = 0;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: use bool

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 14]
```diff
@@ -1776,6 +1776,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+            // Look for trailing spaces
+            while (end < input_len - 1 && isspace((unsigned char)input_str[end + 1])) {
+                end++;
+            }
+
+            // Copy the replacement with removed spaces
+            appendStringInfoString(&output_str, replacement);
+            i = end; // Move the index to the end of the processed sequence
+
+			// skip multiple special characters in a row
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Indentation

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 15]
```diff
@@ -1776,6 +1776,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+*replace_special_chars_FTS(char *input_str) {
+    size_t 		input_len = strlen(input_str);
+    char		*result;
+    char 		*replacement = NULL;
+    char 		*unique_hashes[4];
+    const char 	*special_chars[4] = {"~!&|@#$%^*+=\\;:<>?./", "`", "'", "_"};
+    StringInfoData output_str;
+
+    for (int i = 0; i < 4; i++) {
+        unique_hashes[i] = construct_unique_index_name("specialChars", psprintf("cat%d", i + 1));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Please explain why we have 4 different hashes to replace different kind of characters

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 16]
```diff
@@ -1776,6 +1776,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+            }
+        }
+
+        if (replacement != NULL) {
+            // Remove leading and trailing spaces around special characters
+            size_t start = i;
+            size_t end = i;
+
+            // Look for leading spaces
+            while (start > 0 && isspace((unsigned char)input_str[start - 1])) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why are we doing these?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 17]
```diff
@@ -1776,6 +1776,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+                end++;
+            }
+
+            // Copy the replacement with removed spaces
+            appendStringInfoString(&output_str, replacement);
+            i = end; // Move the index to the end of the processed sequence
+
+			// skip multiple special characters in a row
+			for (int j = 0; j < 4; j++) {
+				while (i < input_len - 1 && strchr(special_chars[j], input_str[i + 1]) != NULL) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: this will not work if characters from different groups can mix, like "@_", we should only skip matching category

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 18]
```diff
@@ -1776,6 +1776,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+        } else {
+            appendStringInfoChar(&output_str, input_str[i]);
+        }
+    }
+
+    // Null-terminate the output string
+    appendStringInfoChar(&output_str, '\0');
+
+    // Duplicate the string before freeing the memory
+    result = pstrdup(output_str.data);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why are we doing this? we can return data itself

File: test/python/expected/sql_validation_framework/expected_create.out [hunk 19]
```diff
@@ -2,6 +2,7 @@ Could not find tests for function assemblyproperty
 Could not find tests for function pltsql_call_handler
 Could not find tests for function pltsql_inline_handler
 Could not find tests for function pltsql_validator
+Could not find tests for function sys.babelfish_fts_rewrite
```
Comment (surendravishnoi) [inline test/python/expected/sql_validation_framework/expected_create.out]: Please add upgrade tests for all C functions added, with calls from function, procedure and view wherever applicable

File: test/JDBC/expected/fts-contains-special-chars.out [hunk 20]
```diff
@@ -0,0 +1,1016 @@
...
+
+
+create fulltext index on test(name) key index test_idx;
+go
+
+select * from test where contains(name, '"one           two"');
+go
+~~START~~
+int#!#text
+1#!#one two three
```
Comment (surendravishnoi) [inline test/JDBC/expected/fts-contains-special-chars.out]: we should think testing in terms of permutation and combination
term1char1,2,3,4,5term2char1,2,3,4,5term3 atleast with three terms and two separators
Now, special characters can mix and match across categories with 1, 2, 3, 4 and 5 characters
line one@_two~`three

File: test/JDBC/expected/fts-contains-special-chars.out [hunk 21]
```diff
@@ -0,0 +1,1016 @@
...
+7#!#one~two
+8#!#one!two
+9#!#one@two
+10#!#one#two
```
Comment (surendravishnoi) [inline test/JDBC/expected/fts-contains-special-chars.out]: we should also test where special characters can at the beginning or end

File: contrib/babelfishpg_tsql/src/fts_parser.y (hunk 7 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/fts_parser.y:158]: This is just a check to initialize the output variable as empty string in case the input string is of `" "` or `""`

File: contrib/babelfishpg_tsql/src/fts_parser.y (hunk 12 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/fts_parser.y]: Done

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 16 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: In order to tackle cases like `"one    :   two"`, in this there are leading and trailing spaces around special character `:` but it still looks for rows with string "one two". So just removing those extra spaces and replacing them with the corresponding unique hashes.

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 22]
```diff
@@ -2590,6 +2592,27 @@ type_name(PG_FUNCTION_ARGS)
...
 
+Datum
+replace_special_chars_fts(PG_FUNCTION_ARGS)
+{
+    text		*input_text = PG_GETARG_TEXT_P(0);
+    char		*input_str = text_to_cstring(input_text);
+	char		*output_str;
+	text 		*result_text;
+
+    // Modify the input_str in place
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/runtime/functions.c]: Nit: Use `/* comment */` for comments.

Comment (basasairohan) [inline contrib/babelfishpg_tsql/runtime/functions.c]: nit : comment style, here and at all other relevant lines

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 23]
```diff
@@ -2590,6 +2592,27 @@ type_name(PG_FUNCTION_ARGS)
...
+    output_str = replace_special_chars_FTS(input_str);
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/runtime/functions.c]: Please distinguish the function name from the above function

**Comment by shalinilohia50** [comment]
Please resolve the merge conflict.

File: test/JDBC/expected/fts-contains-special-chars.out (hunk 20 above)
Comment (roshan0708) [reply test/JDBC/expected/fts-contains-special-chars.out]: We'll be adding unsupported error in such cases.

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 24]
```diff
@@ -113,71 +117,217 @@ simple_term_list:
...
+/* Helper function to generate two strings on the basis of the input string
+ * 1. space separated words
+ * 2. special character (@) separated words
+ *
+ * Case 1: "one two" will generate "one two" and "one@two"
+ * Case 2: "one$two" will generate "one two" and "one@two"
+ */
+static void 
+replaceMultipleSpacesAndSpecialChars(char* input, char **str1, char **str2, bool isEnclosedInQuotes) {
+        size_t inputLen = strlen(input);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: Indentation is 4 spaces

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 25]
```diff
@@ -113,71 +117,217 @@ simple_term_list:
...
+                        if (i != 0 && i != inputLen - 1 && input[i] == '"') {
+                                ereport(ERROR,
+                                    (errcode(ERRCODE_SYNTAX_ERROR),
+                                     errmsg("Syntax error near '%s' in the full-text search condition '%s'.", input + i + 1, input)));
+                        }
+
+                        if (strchr(forbiddenCharsInQuotes, input[i]) != NULL) {
+                                /* Replace forbiddenCharsInQuotes characters based on the position */
+                                while (i + 1 < inputLen && strchr(forbiddenCharsInQuotes, input[i + 1]) != NULL) {
+                                    appendStringInfoString(&modifiedInput, "");
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: not clear why we need to do this. Simply do not add anything as long as we append a '\0' in the end

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 26]
```diff
@@ -113,71 +117,217 @@ simple_term_list:
...
+                /* Null-terminate the modified input string */
+                appendStringInfoChar(&modifiedInput, '\0');
+        } else {
+                appendStringInfoString(&modifiedInput, input);
         }
-        // It's a single word, so no transformation needed
-        return trimmedInputStr;
-    }
+
+        inputLen = strlen(modifiedInput.data);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: modifiedInput has length information already

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 27]
```diff
@@ -1836,6 +1836,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+		} else {
+			appendStringInfoChar(&output_str, input_str[i]);
+		}
+	}
+
+	/* Null-terminate the output string */
+	appendStringInfoChar(&output_str, '\0');
+
+	/* Duplicate the string before freeing the memory */
+    	result = pstrdup(output_str.data);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: you can just return output_str.data, no need to copy it, just remember the pointer

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 28]
```diff
@@ -1836,6 +1836,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+			}
+		}
+
+		if (replacement != NULL) {
+			/* Remove leading and trailing spaces around special characters */
+			size_t start = i;
+			size_t end = i;
+
+			/* Look for leading spaces */
+			while (start > 0 && isspace((unsigned char)input_str[start - 1])) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why are we doing this?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 29]
```diff
@@ -1836,6 +1836,82 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+				end++;
+			}
+
+			/* Copy the replacement with removed spaces */
+			appendStringInfo(&output_str, "-%s-" ,replacement);
+
+			/* Move the index to the end of the processed sequence */
+			i = end; 
+
+			/* skip multiple special characters in a row */
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why are we doing it if we do not support multiple special characters in a row?

File: test/JDBC/expected/fts-contains-vu-verify.out [hunk 30]
```diff
@@ -616,6 +594,1294 @@ GO
...
+
+
+select * from test_special_char_t where contains(name, '"one $%#! two"');
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Concurrent special characters in the full-text search condition are not currently supported in Babelfish)~~
+
+
+select * from test_special_char_t where contains(name, '"one # $ % ^ two"');
```
Comment (surendravishnoi) [inline test/JDBC/expected/fts-contains-vu-verify.out:2366]: what if we remove "", do we still get the error?

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 31]
```diff
@@ -113,71 +117,217 @@ simple_term_list:
...
+                                 errmsg("Syntax error near '%c' in the full-text search condition '%s'.", currentChar, input)));
+                }
+
+                /* Check for consecutive special characters */
+                if (charInSpecialChars != NULL && i + 1 < inputLen && strchr(specialChars, modifiedInput.data[i + 1]) != NULL) {
+                        ereport(ERROR,
+                                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                                 errmsg("Concurrent special characters in the full-text search condition are not currently supported in Babelfish")));
+                } else if (isspace(currentChar) || charInSpecialChars != NULL || (isEnclosedInQuotes && strchr("`'_", currentChar) != NULL)) {
+                        if (!inSpace) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y]: where are we detecting concurrent special characters when space is in-between?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 28 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: This clause basically removes leading and trailing spaces around special characters. Essentially, strings like `"one@two"` and `"one @ two"` these should be treated as same, so that's why adding this clause to trim spaces.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 29 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: We are throwing unsupported error for the search conditions right, but this replace_special_chars function is also applied on the column during the full-text index creation, so if we don't have this clause then let's say a phrase with consecutive special characters is inserted into the table, then it'll crash, so that's why added this clause.

File: test/JDBC/expected/fts-contains-vu-verify.out (hunk 30 above)
Comment (roshan0708) [reply test/JDBC/expected/fts-contains-vu-verify.out:2366]: Without "", only single word is allowed i.e. no spaces in between are allowed, so if we add space it'll automatically throw `syntax error`

File: contrib/babelfishpg_tsql/src/fts_parser.y (hunk 31 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/fts_parser.y]: We are detecting it in the if-clause that is above this for loop. This one:
```
if (strchr(specialChars, input[i]) != NULL) {
        while (i + 1 < inputLen && isspace(input[i + 1])) {
                i++;
        }
        if (i + 1 < inputLen && strchr(specialChars, input[i + 1]) != NULL) {
                ereport(ERROR,
                        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                         errmsg("Concurrent special characters in the full-text search condition are not currently supported in Babelfish")));
        } 
}
appendStringInfoChar(&modifiedInput, input[i]);
```

**Review by basasairohan — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 32]
```diff
@@ -1836,6 +1836,81 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+
+	/* Null-terminate the output string */
+	appendStringInfoChar(&output_str, '\0');
+
+	/* Free the allocated memory */
+	for (int i = 0; i < 4; i++) {
+		pfree(unique_hashes[i]);
+	}
+
+	return  pstrdup(output_str.data);;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: we do not need to create duplicate

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 33]
```diff
@@ -1836,6 +1836,81 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
...
+			/* Check if the current special character is the only one in a sequence */
+			bool is_single = true;
+			size_t next_char_index = i + 1;
+
+			if (next_char_index < input_len) {
+				for (int k = 0; k < 4; k++) {
+					char *is_special_char = strchr(special_chars[k], input_str[next_char_index]);
+					while ((next_char_index < input_len && isspace((unsigned char)input_str[next_char_index])) || (next_char_index < input_len && is_special_char != NULL)) {
+						if (is_special_char != NULL) {
+							is_single = false;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:1876]: we should break out of both loop once is single is false

File: test/JDBC/expected/BABEL-UNSUPPORTED.out [hunk 34]
```diff
@@ -745,9 +745,17 @@ ignore
 
 CREATE DATABASE db_unsupported_ft WITH DEFAULT_FULLTEXT_LANGUAGE = English;
 GO
+~~ERROR (Code: 33557097)~~
```
Comment (surendravishnoi) [inline test/JDBC/expected/BABEL-UNSUPPORTED.out:744]: why this change? earlier we were supporting this for ignore escape hatch but now we give error always

File: test/JDBC/expected/BABEL-UNSUPPORTED.out [hunk 35]
```diff
@@ -719,7 +719,7 @@ CREATE DATABASE db_unsupported_ft WITH DEFAULT_FULLTEXT_LANGUAGE = English;
 GO
 ~~ERROR (Code: 33557097)~~
 
-~~ERROR (Message: 'DEFAULT_FULLTEXT_LANGUAGE' is not currently supported in Babelfish. please use babelfishpg_tsql.escape_hatch_fulltext to ignore)~~
+~~ERROR (Message: 'DEFAULT FULLTEXT LANGUAGE' is not currently supported in Babelfish)~~
```
Comment (surendravishnoi) [inline test/JDBC/expected/BABEL-UNSUPPORTED.out]: why have we changed the error message to not specify escape hatch?

File: test/JDBC/expected/fts-contains-vu-prepare.out [hunk 36]
```diff
@@ -5051,6 +5051,20 @@ GO
 ~~ROW COUNT: 1~~
 
 
+create table test_special_char_t(id int not null, name text);
+go
+
+insert into test_special_char_t values (1, 'one two three'), (2, 'two three one'), (3, 'one'), (4, 'two'), (5, 'three'), (6, 'one`two'), (7, 'one~two'), (8, 'one!two'),(9, 'one@two'), (10, 'one#two'), (11, 'one$two'),(12, 'one%two'), (13, 'one^two'), (14, 'one&two'), (15, 'one*two'), (16, 'one-two'), (17, 'one_two'), (18, 'one+two'), (19, 'one=two'), (20, 'one\two'), (21, 'one|two'), (22, 'one;two'), (23, 'one:two'), (24, 'one''two'), (25, 'one<two'), (26, 'one>two'), (27, 'one.two'), (28, 'one?two'), (29, 'one/two'), (101, 'one :) two'), (100, 'one @ @ @ @ two') ;
```
Comment (surendravishnoi) [inline test/JDBC/expected/fts-contains-vu-prepare.out]: as indicated ealier, we should test with ore permutations/combinations for both inserted data and search string
special chars (0 or more with spaces) term special chars (0 or more with spaces) term    special chars (0 or more with spaces) term special chars (0 or more with spaces)
with and without ""

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 33 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:1876]: Actually I’m trying to skip the group of consecutive special characters all together in a group and copy them in the output string as it is, so that’s why we keep on checking and updating the next_char_index variable and use it accordingly in the below if-else clause

File: test/JDBC/expected/BABEL-UNSUPPORTED.out (hunk 35 above)
Comment (roshan0708) [reply test/JDBC/expected/BABEL-UNSUPPORTED.out]: Will revert this

File: test/JDBC/expected/BABEL-UNSUPPORTED.out (hunk 34 above)
Comment (roshan0708) [reply test/JDBC/expected/BABEL-UNSUPPORTED.out:744]: Will revert this

File: test/JDBC/expected/fts-contains-vu-prepare.out (hunk 36 above)
Comment (roshan0708) [reply test/JDBC/expected/fts-contains-vu-prepare.out]: Done

File: contrib/babelfishpg_tsql/src/fts_parser.y [hunk 37]
```diff
@@ -113,71 +117,210 @@ simple_term_list:
...
+                (errcode(ERRCODE_SYNTAX_ERROR),
+                 errmsg("Syntax error near '%c' in the full-text search condition '%s'.", currentChar, input)));
+        }
+
+        /* Check for consecutive special characters */
+        if (charInSpecialChars != NULL && *(currentCharPtr + 1) && strchr(specialChars, *(currentCharPtr + 1)) != NULL) {
+            ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                 errmsg("Consecutive special characters in the full-text search condition are not currently supported in Babelfish")));
+        } else if (isspace(currentChar) || charInSpecialChars != NULL || (isEnclosedInQuotes && strchr("`'_", currentChar) != NULL)) {
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/fts_parser.y:295]: how do we detect consecutive chars for `'_ etc, we need more test which combine these chars to make sure we cover these cases

**Review by surendravishnoi — APPROVED** [review]
